src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Diagnostics
+{
+    public sealed class ActivitySource : IDisposable
+    {
+        private static object s_syncObject = new object();
+        private static List<ActivitySource> s_activeSources = new List<ActivitySource>();
+        private ActivitySource() { throw new InvalidOperationException(); }
+
+        /// <summary>
+        /// Event to subscribe to get the notification when any ActivitySource object get created or released.
+        /// </summary>
+        public static event EventHandler<ActivitySourceEventArgs>? OperationEvent;
+
+        /// <summary>
+        /// Event to subscribe to get the notification when any Activity object get created or disposed.
+        /// </summary>
+        public event EventHandler<ActivitySourceEventArgs>? ActivityEvent;
+
+        /// <summary>
+        /// Construct an ActivitySource object with the input name
+        /// </summary>
+        /// <param name="name">The name of the ActivitySource object</param>
+        public ActivitySource(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            lock (s_syncObject)
+            {
+                Name = name;
+                s_activeSources.Add(this);
+                EventHandler<ActivitySourceEventArgs>? handlers = OperationEvent;
+                if (handlers != null)
+                {
+                    handlers(this, ActivitySourceEventArgs.s_sourceCreated);
```

Comment:
This is going to invoke arbitrary handlers while holding the static lock.

Reply:
I'll move this outside the lock. Thanks for pointing at that.

---
In reply to: [383014742](https://github.com/dotnet/runtime/pull/32660#discussion_r383014742) [](ancestors = 383014742)